src/Lorentz/Contracts/RegistryDAO/Types.hs

```diff
@@ -0,0 +1,162 @@
+  , ceSlashDivisionValue :: Natural
+  , ceMaxProposalSize :: Natural
+  }
+  deriving stock (Generic)
+
+instance (NiceComparable k, Ord k, IsoValue k, IsoValue v)
+  => (IsoValue (RegistryDaoContractExtra k v))
+
+instance (HasAnnotation k, HasAnnotation v)
+  => HasAnnotation (RegistryDaoContractExtra k v) where
+    annOptions = DAO.baseDaoAnnOptions
+
+instance (Ord k, TypeHasDoc k, TypeHasDoc v)
+  => TypeHasDoc (RegistryDaoContractExtra k v) where
+    typeDocMdDescription =
+      "TODO"
+    typeDocMdReference = poly2TypeDocMdReference
+    typeDocHaskellRep = concreteTypeDocHaskellRep @(RegistryDaoContractExtra ByteString ByteString)
+    typeDocMichelsonRep = concreteTypeDocMichelsonRep @(RegistryDaoContractExtra ByteString ByteString)
+
+instance Default (RegistryDaoContractExtra k v) where
+  def = RegistryDaoContractExtra
+    { ceRegistry = def
+    , ceFrozenScaleValue = 1
+    , ceFrozenExtraValue = 0
+    , ceSlashScaleValue = 1
+    , ceSlashDivisionValue = 1
+    , ceMaxProposalSize = 100
+    }
+
+instance IsoRegistryDaoProposalMetadata k v =>
+    StoreHasSubmap (RegistryDaoContractExtra k v) "ceRegistry" k (RegistryEntry v) where
+  storeSubmapOps = storeSubmapOpsDeeper #ceRegistry
+
+
+data RegistryEntry v = RegistryEntry
+  { reValue :: Maybe v
+   -- optional statistics on given proposal
+  , reLastAgoraPostId :: Natural
+  , reLastUpdated :: Timestamp
+  } deriving stock (Generic)
+  deriving anyclass (IsoValue)
+
+instance (HasAnnotation v)
+  => HasAnnotation (RegistryEntry v) where
+    annOptions = DAO.baseDaoAnnOptions
+
+instance (TypeHasDoc v)
+  => TypeHasDoc (RegistryEntry v) where
+    typeDocMdDescription =
+      "TODO"
+    typeDocMdReference = poly1TypeDocMdReference
+    typeDocHaskellRep = concreteTypeDocHaskellRep @(RegistryEntry ByteString)
+    typeDocMichelsonRep = concreteTypeDocMichelsonRep @(RegistryEntry ByteString)
+
+
+data RegistryDaoProposalMetadata k v
+  = NormalProposalType (NormalProposal k v)
+  | ConfigProposalType ConfigProposal
+  deriving stock (Generic)
+
+type IsoRegistryDaoProposalMetadata k v = (NiceComparable k, Ord k, IsoValue k, KnownValue v)
+
+instance IsoRegistryDaoProposalMetadata k v
+  => (IsoValue (RegistryDaoProposalMetadata k v))
+
+instance (HasAnnotation k, HasAnnotation v)
+  => HasAnnotation (RegistryDaoProposalMetadata k v) where
+    annOptions = DAO.baseDaoAnnOptions
+
+instance (TypeHasDoc k, TypeHasDoc v)
+  => TypeHasDoc (RegistryDaoProposalMetadata k v) where
+    typeDocMdDescription =
+      "TODO"
+    typeDocMdReference = poly2TypeDocMdReference
+    typeDocHaskellRep = concreteTypeDocHaskellRep @(RegistryDaoProposalMetadata ByteString ByteString)
+    typeDocMichelsonRep = concreteTypeDocMichelsonRep @(RegistryDaoProposalMetadata ByteString ByteString)
+
+
+-- | A registry proposal. It will update registry list in contract storage when got accepted.
+data NormalProposal k v = NormalProposal
+  { npAgoraPostId :: Natural
```

Comment:
I would very much prefer to wrap this `Natural` into a newtype because we should never use it as a number, i. e. never apply arithmetic operations to it. Wrapping it into a newtype can prevent us from doing that.

Reply:
Sure, I've updated this to use a new type now.